src/storage/v2/property_value.hpp

```diff
@@ -31,73 +33,176 @@ class PropertyValueException : public utils::BasicException {
+      : alloc_{alloc}, list_v{.val_ = list_t{std::move(value), alloc}} {}
+
   /// @throw std::bad_alloc
-  explicit PropertyValue(std::map<std::string, PropertyValue> value) : map_v{.val_ = std::move(value)} {}
+  explicit PropertyValueImpl(map_t const &value) : alloc_{value.get_allocator()}, map_v{.val_ = value} {}
+  explicit PropertyValueImpl(map_t &&value) : alloc_{value.get_allocator()}, map_v{.val_ = std::move(value)} {}
+  explicit PropertyValueImpl(map_t const &value, allocator_type const &alloc)
+      : alloc_{alloc}, map_v{.val_ = map_t{value, alloc}} {}
+  explicit PropertyValueImpl(map_t &&value, allocator_type const &alloc)
+      : alloc_{alloc}, map_v{.val_ = map_t{std::move(value), alloc}} {}
 
   // copy constructor
   /// @throw std::bad_alloc
-  PropertyValue(const PropertyValue &other);
+  PropertyValueImpl(const PropertyValueImpl &other);
+  PropertyValueImpl(const PropertyValueImpl &other, allocator_type const &alloc);
 
   // move constructor
-  PropertyValue(PropertyValue &&other) noexcept;
+  PropertyValueImpl(PropertyValueImpl &&other) noexcept;
+  PropertyValueImpl(PropertyValueImpl &&other, allocator_type const &alloc) noexcept;
 
   // copy assignment
   /// @throw std::bad_alloc
-  PropertyValue &operator=(const PropertyValue &other);
+  PropertyValueImpl &operator=(const PropertyValueImpl &other);
 
   // move assignment
-  PropertyValue &operator=(PropertyValue &&other) noexcept;
+  PropertyValueImpl &operator=(PropertyValueImpl &&other) noexcept(
+      std::allocator_traits<Alloc>::is_always_equal::value ||
+      std::allocator_traits<Alloc>::propagate_on_container_move_assignment::value);
+
   // TODO: Implement copy assignment operators for primitive types.
   // TODO: Implement copy and move assignment operators for non-primitive types.
 
-  ~PropertyValue() {
+  template <typename Alloc2>
+  friend class PropertyValueImpl;
+
+  /// Copy accross allocators
+  template <typename Alloc2>
+  requires(!std::same_as<Alloc, Alloc2>)
+      PropertyValueImpl(PropertyValueImpl<Alloc2> const &other, allocator_type const &alloc = allocator_type{})
+      : alloc_{alloc}, type_{other.type_} {
+    switch (other.type_) {
+      case Type::Null:
+        return;
+      case Type::Bool:
+        bool_v.val_ = other.bool_v.val_;
+        return;
+      case Type::Int:
+        int_v.val_ = other.int_v.val_;
+        return;
+      case Type::Double:
+        double_v.val_ = other.double_v.val_;
+        return;
+      case Type::String:
+        std::allocator_traits<allocator_type>::construct(alloc_, &string_v.val_, other.string_v.val_);
+        return;
+      case Type::List: {
+        auto converted_val = list_t(alloc_);
+        converted_val.reserve(other.list_v.val_.size());
+        for (auto const &val : other.list_v.val_) {
+          converted_val.emplace_back(val);
+        }
+        std::allocator_traits<allocator_type>::construct(alloc_, &list_v.val_, std::move(converted_val));
+        return;
+      }
```

Comment:
1. Vector's assignment `operator=` does not support vectors of different allocators. https://en.cppreference.com/w/cpp/container/vector/operator%3D
2. We need to, in the interim handle, `std::allocator` <-> `std::pmr::polymorphic_allocator`. Which makes `std::vector<T, std::allocator>` incompatible with `std::vector<T, std::pmr::polymorphic_allocator>`
3. Here we can't do assignment, because `list_v.val_` is not an object yet...its lifetime has not started.
4. Having said that we can do `alloc_trait::construct(alloc_, &list_v.val_, other.list_v.val_.begin(), other.list_v.val_.end());` which is iterator construction with a provided allocator.